.. _rnaseqqcpipeline:
.. _rnaseqqcpipeline:

==============
3' Bias
==============

Summary
=======

Aims
----
This report generates a barplot for each sample showing the normalised
read depth across all transcripts for the first 3000 base pairs upstream
of the 3' end.  This is used to identify any strong 3' bias, which
can occur with polyA enriched RNA-seq.  Idenfication of this bias
allows it to be taken into account in later analysis.

Inputs
------
The input is a database table, threeprimebiasprofiles, generated by bam2geneprofile
script.  Aggregated over all the gene models in the genesets gtf file provided by the user, this table
contains the number of reads at each position, starting from the 3' end of the transcript model
and cropped at 3000 base pairs upstream.  Transcripts shorter than 3000 base pairs are excluded.

Outputs
-------
The report generates a bar plot for each sample.  Each bin on the bar plot is a
single position.  Starting from the 3' end at the right hand side of the x axis, the
normalised, aggregated read count at each position is shown.

How the results are generated
-----------------------------
Bam2geneprofile performs the following steps:

- Builds a theoretical mRNA transcript for each gene model specified in the gene set gtf by stitching exons into one piece of mRNA
- Aggregates these into a single "metagene" object
- Calculates the aggregated read count at each position upstream of the 3' end of this object (starting at the TSS)
- Combines these into a table for each sample, where each row is a bin representing one position relative to the 3' end
- Concatenates these tables

What you should expect from results
-----------------------------------
- Ideally, there distribution should be flat, with no increase in read depth towards the 3' end
- However, with polyA enriched RNA seq a bias is likely to be noticable (see Table 1 in this publication http://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-15-419)
- cDNA fragmentation can also lead to 3' bias (see Figure 3 of this publication http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2949280/figure/F3/)
- If there is a very marked underrepresentation of the 5' ends, this may indicate an issue with the library preperation.
- Standard differential expression analysis tools do not take into account whereabouts in the gene each read falls.
- However, with strong three prime bias shorter genes will appear overexpressed, as the low coverage at the 5' end will be less pronounced
- More information is available at http://seqanswers.com/forums/showthread.php?t=9839 and http://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-290


Examples
--------
The good

.. report:: GoodExample.Tracker
   :render: myRenderer
   :transform: myTransform
   :options: myAesthetics

Ideally, these plots should show an even read depth along the read.


The bad

.. report:: BadExample.Tracker
   :render: myRenderer
   :transform: myTransform
   :options: myAesthetics

This plot shows data with a strong 3' bias.

More bad examples `<http://myBadData.html >`

Your data:

.. report:: RnaseqqcReport.ThreePrimeBias
   :render: r-ggplot
   :statement: aes(bin, counts) +
   	       geom_bar(size=0.3, stat="identity",
   	       fill="springgreen4", colour="springgreen4") +
	       scale_color_brewer(palette = "Set1") +
	       theme(panel.background=element_rect(fill="white",
			colour="white"), legend.position="none") +
	       ggtitle("3 Prime Bias") +
	       scale_x_discrete(name="Distance from 3 prime end",
	       			labels=c(-3000, -2500, -2000, -1500,
   						-1000, -500, "3\nprime\nend"),
   	       breaks=c(0, 500, 1000, 1500, 2000, 2500, 3000)) +
	       ylab("Normalised Read Count")


Commentary
  This will take the form of some active comments.  This will require the report to
  be published so that it is hosted on the CGAT server/ comments on the DISQUS server.

