.. _strandbias:

===========
Strand Bias
===========

Summary
=======

If the RNA-Seq is stranded, we would expect the vast majority of the
reads to come from the sense strand of annotated transcripts. Here we
inspect the ratio of sense to antisense reads.


Results
=======

.. report:: RnaseqqcReport.PicardStrandBias
   :render: r-ggplot
   :statement: aes(track, value) +
	       geom_bar(stat="identity") +
	       theme_bw()

.. report:: RnaseqqcReport.PicardThreePrimeBias
   :render: r-ggplot
   :split-at: 10
   :statement: aes(track, value) +
	       geom_bar(stat="identity") +
	       theme_bw()
	       

Aims
----
If the RNA-Seq is stranded, we would expect the vast majority of the
reads to come from the sense strand of annotated transcripts. Here we
inspect the ratio of sense to antisense reads.
 
Inputs
------
The input is a database table, picard_rna_metrics, generated by Picard RnaSeqMetrics


Outputs
-------
The report generates a bar plot for each sample


Examples
--------
The good

.. report:: GoodExample.Tracker
   :render: myRenderer
   :transform: myTransform
   :options: myAesthetics

Ideally, these plots should show an even read depth along the read.


The bad

.. report:: BadExample.Tracker
   :render: myRenderer
   :transform: myTransform
   :options: myAesthetics

This plot shows data with a strong 3' bias.

More bad examples `<http://myBadData.html >`


Commentary
  This will take the form of some active comments.  This will require the report to
  be published so that it is hosted on the CGAT server/ comments on the DISQUS server.
