.. _heatmap:

=======
Heatmap
=======

Summary
=======

Samples similarity can be measured in a number of ways.  Primarily we
are interested in whether experimental samples are more similar to
their biological replicates _within_ a condition compared to between
conditions.  Deviation from the expected groupings may be indicative
of batch effects, experimental artifacts and biases, sample mix-up or
experimental failure.

Results
=======
.. report:: RnaseqqcReport.SampleHeatmap
   :render: RnaseqqcReport.TranscriptQuantificationHeatmap

   Similarity of samples using pair-wise Pearson correlations.
   Samples are clustered by average linkaged hierarchical clustering.

About this section
==================

Inputs
------

A table of expression values transcripts per million (TPM) for each
transcript and sample generated by Sailfish and stored in a table in
the pipeline database

Outputs
-------

A heatmap with row and column dendrograms of the pairwise sample
Pearson correlation coefficients.  A colour sidebar is added to
demonstrate different aspects of the experimental design, including
replicates and conditions.

How the results are generated
-----------------------------

1. Expression results are organised as transcripts down the rows and
   samples across the columns

2. This matrix is transposed and the pairwise Pearson correlation
   coefficients are computed between these transposed rows (samples)

What you should expect from the results
---------------------------------------

Samples that are experimentally closer together, i.e. replicates
within an experimental condition, should lie closer together in the
heatmap.  Samples from different conditions should lie distal to each
other.  If samples cluster unexpectedly this may indicate confounding
from additional sources, such as library preparation, sequencing lane
or other experimental factors.  Currently the colour bar only
represents the different replicate groups (NB: will add `condition`
later when these have been set as required fields).
